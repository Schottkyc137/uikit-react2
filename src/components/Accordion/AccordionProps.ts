import { HTMLProps } from 'react';

export interface AccordionProps extends HTMLProps<HTMLUListElement> {
  /**
   * Index of the element to open initially.
   *
   * @default false
   */
  active?: number | false;

  /**
   * Reveal item directly or with a transition.
   *
   * @default true
   */
  animation?: boolean;

  /**
   * Allow all items to be closed
   *
   * @default true
   */
  collapsible?: boolean;

  /**
   * The content selector, which selects the accordion content elements.
   *
   * @default '> .uk-accordion-content'
   */
  content?: string;

  /**
   * Animation duration in milliseconds.
   *
   * @default 200
   */
  duration?: number;

  /**
   * Allow multiple open items.
   *
   * @default false
   */
  multiple?: boolean;

  /**
   * CSS selector of the element(s) to toggle.
   *
   * @default '> *'
   */
  targets?: string;

  /**
   * The toggle selector, which toggles accordion items.
   *
   * @default '> .uk-accordion-title'
   */
  toggle?: string;

  /**
   * The transition to use when revealing items. Use keyword for easing functions.
   *
   * @default 'ease'
   */
  transition?: string;

  /**
   * Pixel offset added to scroll top.
   *
   * @default 0
   */
  offset?: number;
}

export interface AccordionItemProps extends HTMLProps<HTMLDivElement> {
  /**
   * whether this should be open initially.
   * You can also instead specify the active element of the parent accordion
   */
  open?: boolean;

  /**
   * The title of the item
   */
  title?: string;
}
