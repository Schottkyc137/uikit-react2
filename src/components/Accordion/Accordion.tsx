import React from 'react';
import { AccordionProps } from './AccordionProps';
import UIkit from 'uikit';

/**
 * Create a list of items that can be shown individually by clicking an item's header.
 * You create an accordion by using the `Accordion` as parent element
 * with one or more `AccordionItem`s as children. These children can
 * have arbitrary content.
 * <pre>{@code
 * <Accordion>
 *   <AccordionItem>...</AccordionItem>
 *   <AccordionItem>...</AccordionItem>
 * </Accordion>
 * }</pre>
 */
export class Accordion extends React.Component<AccordionProps> {
  private el: HTMLUListElement | null;
  private accordion: UIkit.UIkitAccordionElement | null;

  componentDidMount() {
    if (this.el !== null) {
      this.accordion = UIkit.accordion(this.el, {
        //@ts-ignore
        active: this.props.active,
        animation: this.props.animation,
        collapsible: this.props.collapsible,
        content: this.props.content,
        duration: this.props.duration,
        multiple: this.props.multiple,
        targets: this.props.targets,
        toggle: this.props.toggle,
        transition: this.props.transition,
        offset: this.props.offset,
      });
    }
  }

  /**
   * Toggles the content pane.
   * @param index Accordion pane to toggle. 0 based index.
   * @param animate Suppress opening animation by passing false.
   */
  public toggle(index: number, animate: boolean) {
    this.accordion?.toggle(index, animate);
  }

  render() {
    const {
      active,
      animation,
      collapsible,
      content,
      duration,
      multiple,
      targets,
      toggle,
      transition,
      offset,
      ...rest
    } = this.props;
    return (
      <ul data-testid="accordion" ref={(el) => (this.el = el)} {...rest} />
    );
  }
}
