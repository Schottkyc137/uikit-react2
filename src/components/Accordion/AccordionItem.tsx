import React from 'react';
import { AccordionItemProps } from './AccordionProps';
import { extendClasses } from '../util/class-names';

/**
 * The container class for an accordion item.
 *
 * The Props that are specific to the HTML-Element apply to the container
 * of the content
 * @param props
 * @constructor
 */
export function AccordionItem(props: AccordionItemProps) {
  const { className, title, open, ...rest } = props;

  return (
    <li className={open ? 'uk-open' : undefined}>
      <a className="uk-accordion-title" href="/#">
        {props.title}
      </a>
      <div
        className={extendClasses(className, 'uk-accordion-content')}
        {...rest}
      >
        {props.children}
      </div>
    </li>
  );
}
