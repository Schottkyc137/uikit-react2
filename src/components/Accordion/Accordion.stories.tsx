import React from 'react';
import { Accordion } from './Accordion';
import { AccordionItem } from './AccordionItem';

export default {
  title: 'Accordion',
  component: Accordion,
};

export const Basic = () => (
  <Accordion>
    <AccordionItem title="Item 1">
      <p>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
        tempor incididunt ut labore et dolore magna aliqua.
      </p>
    </AccordionItem>
    <AccordionItem title="Item 2">
      <p>
        Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi
        ut aliquip ex ea commodo consequat. Duis aute irure dolor reprehenderit.
      </p>
    </AccordionItem>
    <AccordionItem title="Item 3">
      <p>
        Duis aute irure dolor in reprehenderit in voluptate velit esse cillum
        dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat
        proident.
      </p>
    </AccordionItem>
  </Accordion>
);

export const NoCollapsing = () => (
  <Accordion collapsible={false}>
    <AccordionItem title="Item 1">
      <p>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
        tempor incididunt ut labore et dolore magna aliqua.
      </p>
    </AccordionItem>
    <AccordionItem title="Item 2">
      <p>
        Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi
        ut aliquip ex ea commodo consequat. Duis aute irure dolor reprehenderit.
      </p>
    </AccordionItem>
    <AccordionItem title="Item 3">
      <p>
        Duis aute irure dolor in reprehenderit in voluptate velit esse cillum
        dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat
        proident.
      </p>
    </AccordionItem>
  </Accordion>
);

export const Multiple = () => (
  <Accordion multiple={true}>
    <AccordionItem title="Item 1">
      <p>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
        tempor incididunt ut labore et dolore magna aliqua.
      </p>
    </AccordionItem>
    <AccordionItem title="Item 2">
      <p>
        Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi
        ut aliquip ex ea commodo consequat. Duis aute irure dolor reprehenderit.
      </p>
    </AccordionItem>
    <AccordionItem title="Item 3">
      <p>
        Duis aute irure dolor in reprehenderit in voluptate velit esse cillum
        dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat
        proident.
      </p>
    </AccordionItem>
  </Accordion>
);
