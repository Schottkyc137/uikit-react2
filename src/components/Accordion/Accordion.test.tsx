import { render, screen } from '@testing-library/react';
import React from 'react';
import { Accordion } from './Accordion';
import { AccordionItem } from './AccordionItem';
import 'intersection-observer';

describe('Accordion', () => {
  it('renders without crashing', () => {
    expect(render(<Accordion />)).toBeTruthy();
  });

  it('renders a snapshot', () => {
    render(
      <Accordion>
        <AccordionItem open>Panel</AccordionItem>
      </Accordion>
    );
    expect(screen.getByTestId('accordion')).toMatchSnapshot();
  });
});
