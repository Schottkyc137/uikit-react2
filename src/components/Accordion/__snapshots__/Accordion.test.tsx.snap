// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion renders a snapshot 1`] = `
<ul
  class="uk-accordion"
  data-testid="accordion"
>
  <li
    class="uk-open"
  >
    <a
      class="uk-accordion-title"
      href="/#"
    />
    <div
      class="uk-accordion-content"
    >
      Panel
    </div>
  </li>
</ul>
`;
